lib/src/ConfigLoader.cc

```diff
@@ -570,16 +570,28 @@ static void loadListeners(const Json::Value &listeners)
         auto key = listener.get("key", "").asString();
         auto useOldTLS = listener.get("use_old_tls", false).asBool();
         LOG_TRACE << "Add listener:" << addr << ":" << port;
-        drogon::app().addListener(addr, port, useSSL, cert, key, useOldTLS);
+        std::vector<std::pair<std::string, std::string>> conf;
+        for (const auto &opt : listener["ssl_conf"])
+        {
+            conf.emplace_back(opt[0].asString(), opt[1].asString());
```

Comment:
Please reject malformed `opt`. There must be 1 or two elements and all elements must be a string. Otherwise people write a large option and be wondering why they aren't applied.

Reply:
I'm not that familiar with how usually drogon did, so here is some questions.

For malformed `opt`, do we need to check the `opt` with OpenSSL's `SSL_CONF_cmd_value_type` ?
For example, we could check `SSL_CONF_TYPE_UNKNOWN` on `opt[0]` and `SSL_CONF_TYPE_NONE` on `opt[1]`?
Or maybe it is enough to do something like below?

```cpp
if (opt.size() != 1 and opt.size() != 2) {
  std::cerr << "Invalid ssl_conf provided: " << opt << std::endl;
  exit(1);
  // or use LOG_ERROR, and simply skip the invalid opt
}
```

Reply:
Just do the basic check in Drogon. (And _if you want to_, check `SSL_CONF_cmd_value_type`, print, and shutdown the connection in trantor). 

```
if (opt.size() == 0 || opt.size() > 2) 
{
    LOG_FATAL << "SSL configuration option should be an 1 or 2-element array";
    abort();
}
if(.....)
```

@an-tao IMO this is a fatal error. I wouldn't want a misconfigured server (ex. using the wrong DH curve) to be running. What do you think?

Reply:
After some thought, maybe we could only allow `opt` with size equals to 2?
Reason: If someone is going to use this feature, he/she probably has read the [document in openssl](https://www.openssl.org/docs/manmaster/man3/SSL_CONF_cmd.html). Thus, they would know it should be an (option, value) pair even though the value might be ignored. 

WDYT, @marty1885 ?

Reply:
IMO it's annoying to have restrictions in the configuration that have nothing to do with semantic (if I'm the user). From experience, that kind of bugs takes forever to debug and is just bad user experience.

Reply:
Another stupid question. Are there any difference using `std::cerr` or `LOG_FATAL` in drogon (i.e. in some cases we should use `LOG_FATAL` and vice versa)?
I found that in `lib/src/ConfigLoader.cc` all the rejection uses `std::cerr` and `exit`.

Reply:
`LOG_*` optionally logs to file and automatically appends timestamps and allows filtering of messages. I think we use cerr in the config loader because the config includes the filter and log file settings. 

@an-tao ?